### ISSUE TITLE
`.timeout(ms, {cancel: true})` not working for transactions

### ISSUE THREAD

**Issue by martinmacko47** [knex/knex#2211]
This issue applies only to mysql and mariadb as `.timeout(ms, {cancel: true})` is implemented for them only. Tested with knex 0.13.0

If **in a transaction**, `client.acquireConnection` acquires always the same connection. Therefore, [`const acquiringConn = this.acquireConnection()`](https://github.com/tgriesser/knex/blob/8f7d2e297799d4b70de726b8cfc9d4939fe3923b/src/dialects/mysql/index.js#L148) in `cancelQuery` returns the same connection as is being canceled (`connectionToKill.threadId` and `acquiringConn.threadId` are the same).

As the result the query `KILL QUERY ?` (with `?` bound to `connectionToKill.threadId`) is executed only **after** the original query finishes, as they both use the same connection. The original query is not killed at all, and only after it finishes the kill query just kills itself, as it has the same `threadId`.

**How to reproduce the bug**
```
knex.transaction(function(trx) {
  return trx.select(trx.raw('SLEEP(47)')).timeout(5000, {cancel: true})
}).then(function(res){
  console.log('OK', res)
}).catch(function(err){
  console.log('ERROR', err)
})
```

*Result*: The query runs for 47 seconds (check with `SHOW PROCESSLIST`) and then (after the 47 seconds) throws error: "Error: After query timeout of 5000ms exceeded, cancelling of query failed."

*Expected result*: The query should be killed after 5 seconds and should throw error: "TimeoutError: Defined query timeout of 5000ms exceeded when running query."

If the query is not executed in a transaction (e.g. `knex.select(knex.raw('SLEEP(47)')).timeout(5000, {cancel: true})`) it is killed as expected.

**Possible fix**

Change [`const acquiringConn = this.acquireConnection()`](https://github.com/tgriesser/knex/blob/8f7d2e297799d4b70de726b8cfc9d4939fe3923b/src/dialects/mysql/index.js#L148) in `cancelQuery` to always get a new connection, even if in a transaction.

**Comment by deje1011** [comment]
I am facing the same problem with PostgreSQL. Does anyone have a workaround?

**Comment by martinmacko47** [comment]
@kibertoad I've found another case when cancelling timeouted queries in a transaction fails on MySQL:

```
knex.transaction(function(trx) {
  return Promise.all([
    firstSlowQuery(trx).timeout(5000, {cancel: true}),
    secondSlowQuery(trx).timeout(5000, {cancel: true}),
    ...
    lastSlowQuery(trx).timeout(5000, {cancel: true}),
  ])
})
```

As the queries are in the transaction, they get queued in the driver which sends them to DB one by one. As all queries will timeout at approximately the same time, all their `KILL QUERY ?` commands will be also sent to DB at approximately the same time. Therefore they will kill only the first query. After that the driver sends the second query to DB which will be never killed, as its timeout already elapsed and its `KILL QUERY ?` command was already sent.

The problem occurs only on MySQL. Postgres marks the transaction failed after killing the query, therefore any further queries in the same transaction will immediately fail.

I guess we should simulate the same behaviour on MySQL as well. Either by removing the timeouted queries from the driver queue, or by rolling back the transaction with the killed timeouted query.

I've written a test to replicate this bug. It passes for Postgres and fails for MySQL: https://github.com/martinmacko47/knex/pull/4/files#diff-ef380392f7c99984959538380cd3a01631d49da7eb768b02fd71695bfb2a3febR981-R1052

Should I open a new issue for this, or reopen this one?

**Comment by elhigu** [comment]
Actually also in postgresql it is configurable behaviour if transaction should fail when one of the queries sent to it fails. Is it possible to make mysql to rollback transaction implicitly in that case too? 

I'm not sure what actually should be done in this case... Maybe it is better to open new issue for this.

**Comment by martinmacko47** [comment]
@elhigu Ok. I opened a new issue for it: #4415
